### PR TITLE
d2-vscode-extension

### PR DESCRIPTION
**Pull request by BarryNolte**
First PR creating the viewer for d2 files in VsCode.

#4 - For Reference

Fixes:
#9 and #13

### CODE REVIEW

**Comment by BarryNolte** [comment]
@alixander It’s as ready as it’ll ever be. Only prerequisite needed is to have d2 on your path.

File: package.json
Comment (gavin-ts) [inline package.json:53]: was the `themes` section deleted accidentally?

File: .vscode/launch.json
Comment (gavin-ts) [inline .vscode/launch.json:4]: I think comments are technically invalid in json so we might want to remove these

File: src/docToPreviewGenerator.ts
Comment (gavin-ts) [inline src/docToPreviewGenerator.ts:89]: we probably want to use a separate function to send an error message and we can handle that separately from updating the SVG

File: src/browserWindow.ts
Comment (gavin-ts) [inline src/browserWindow.ts:8]: can we define this in a `.html` likewise for the css and js?

File: src/browserWindow.ts
Comment (gavin-ts) [inline src/browserWindow.ts:97]: I think a zoomFit would be better for the default onload.

File: src/extension.ts
Comment (gavin-ts) [inline src/extension.ts:30]: do we need the refreshTimer, can't we just run each time this TextDocumentChangeEvent is fired (possibly with a debounce if it is too frequent)?

**Review by gavin-ts — CHANGES_REQUESTED** [review]
Thanks for setting this up! I added a bunch of notes, but some of these are larger issues to tackle so I was thinking we could merge this into the `dev` branch after some of the smaller changes and handle the rest in their own PRs?

Looks good with the fit to screen option:
Screen Shot 2022-12-21 at 2 50 35 PM

File: src/extension.ts
Comment (BarryNolte) [reply src/extension.ts:30]: The timer is the debounce, if a change is made, then another one isn't detected for the given time (1.5 sec by default), then update occurs.  The reset function just says, "I know your counting, but start over, there are more changes coming"

File: src/docToPreviewGenerator.ts
Comment (BarryNolte) [reply src/docToPreviewGenerator.ts:89]: Yes.  I was way too clever here, one of those strings will be empty, so it will either display an error or the svg.  I'll make this more sane.

File: .vscode/launch.json
Comment (BarryNolte) [reply .vscode/launch.json:4]: This file, at least originally, is vscode generated, so I'm inclined to say, if they don't mind comments, I don't either.

File: package.json
Comment (BarryNolte) [reply package.json:53]: On purpose, it was causing problems since it wasn't fully functional and it's meant for full blown VSCode themes, which I didn't think was the intention. If you want the preview to follow the current theme, that's doable, but is a larger work item and may involve some coordination with the d2 exe.

File: src/browserWindow.ts
Comment (BarryNolte) [reply src/browserWindow.ts:97]: Agree, I'll change this.

File: src/extension.ts
Comment (gavin-ts) [reply src/extension.ts:30]: hmm okay that makes sense I think we can use `setTimeout` for the debounce instead of `setInterval` and then we shouldn't need to call `this.stop()` and can just call `this.callback()` since it would only run once by default. It makes it clearer to me that it isn't going to be run at a regular interval which is how I misread it initially.

File: src/browserWindow.ts
Comment (BarryNolte) [reply src/browserWindow.ts:168]: I'll fix this, it's better all around.

File: src/browserWindow.ts
Comment (BarryNolte) [reply src/browserWindow.ts]: Change [9650284](https://github.com/terrastruct/d2-vscode/pull/11/commits/9650284237a5893b5c8808b2b4bdee7254f62b04) fixes this.

File: src/browserWindow.ts
Comment (BarryNolte) [reply src/browserWindow.ts]: I changed this to load an HTML file, but breaking up into html/css/js file causes security issues, which is solved by using asWebViewUri to get an acceptable Uri to the css/js files, but then has to be inserted into the html text, which gets us back to the original problem of having the html defined as strings in the code.

File: src/browserWindow.ts
Comment (BarryNolte) [reply src/browserWindow.ts]: II changed this to load an HTML file, but breaking up into html/css/js file causes security issues, which is solved by using asWebViewUri to get an acceptable Uri to the css/js files, but then has to be inserted into the html text, which gets us back to the original problem of having the html defined as strings in the code.

**Comment by BarryNolte** [comment]
@gavin-ts All requested changes made in commit: [9650284](https://github.com/terrastruct/d2-vscode/pull/11/commits/9650284237a5893b5c8808b2b4bdee7254f62b04)

File: package.json
Comment (aaronjheng) [inline package.json:3]: Should be "D2".

File: .vscode/launch.json
Comment (gavin-ts) [inline .vscode/launch.json:33]: we no longer need this right?

File: package.json
Comment (gavin-ts) [inline package.json:92]:
```suggestion
          "description": "Automatically update on saving the d2 file.  Turn off to avoid excess updates when Auto Update is enabled."
```

I think we want to have the save event stop the refresh timer if this is on, so we won't need this second sentence and it will work well if both settings are enabled.

File: package.json
Comment (BarryNolte) [reply package.json]: Changed.  But... Can I get some guidance from Terrastruct on proper branding?  Proper casing of names, colors, etc...

File: package.json
Comment (gavin-ts) [inline package.json:82]:
```suggestion
          "description": "Automatically update the preview as you edit the d2 file."
```

File: package.json
Comment (gavin-ts) [inline package.json:87]:
```suggestion
          "description": "Specifies the amount of time (in ms) to wait between keystrokes before updating the preview."
```

File: package.json
Comment (gavin-ts) [inline package.json:99]:
```suggestion
          "command": "D2.ShowPreviewWindow",
```
maybe just D2 for these?

File: package.json
Comment (gavin-ts) [inline package.json:139]: do we need this for the non-test ts files?

File: pages/previewPage.html
Comment (gavin-ts) [inline pages/previewPage.html:114]:
```suggestion
```

File: pages/previewPage.html
Comment (gavin-ts) [inline pages/previewPage.html:104]: I think in the future we will want to be able to also update the zoom level with `fit` so you can fit and then zoom in or out relative to that. This can be a follow up issue though.

File: src/extension.ts
Comment (gavin-ts) [inline src/extension.ts:43]: here we can add the part that checks if autoUpdate is also enabled and we can stop the timer if so

**Review by gavin-ts — CHANGES_REQUESTED** [review]
nice we will want to add a note to the PR body that this fixes #9 and #13

File: .vscode/launch.json
Comment (BarryNolte) [reply .vscode/launch.json:33]: Very much need that for generating the extension within vscode and debugging

File: pages/previewPage.html
Comment (BarryNolte) [reply pages/previewPage.html:104]: Yes, I want that too.

File: src/extension.ts
Comment (BarryNolte) [reply src/extension.ts:55]: Yup.  Just checked, its fixed in PR #15.

File: package.json
Comment (BarryNolte) [reply package.json:139]: I'll strip out all the test stuff since it doesn't do anything, the other ares needed for development work.

File: package.json
Comment (BarryNolte) [reply package.json:92]: Fixed in PR #16

File: package.json
Comment (BarryNolte) [reply package.json:99]: I was afraid this would cause name collisions, but it's scoped to the extension.  Fixed in PR #16.

File: package.json
Comment (BarryNolte) [reply package.json:139]: Removed all the unneeded tasks in PR #16

File: pages/previewPage.html
Comment (BarryNolte) [reply pages/previewPage.html:104]: Fixed in PR #16.  How dare you make me do math...

File: src/extension.ts
Comment (BarryNolte) [reply src/extension.ts:43]: Fixed in PR #16.  I changed some of the logic to take care of 'not' updating on a autosave, if we have update on save turned off.

File: src/extension.ts
Comment (gavin-ts) [reply src/extension.ts:55]: can we add this to this PR?

File: package.json
Comment (gavin-ts) [reply package.json:139]: can we add to this PR?

File: package.json
Comment (gavin-ts) [reply package.json:99]: this PR?

**Review by gavin-ts — CHANGES_REQUESTED** [review]
I think it should be good to merge with these changes

File: package.json
Comment (BarryNolte) [reply package.json:99]: The next PR.  This one is getting a little out of hand.

File: package.json
Comment (BarryNolte) [reply package.json:139]: Done

File: src/extension.ts
Comment (BarryNolte) [reply src/extension.ts:55]: Fixed in this PR.

Comment (BarryNolte) [reply src/extension.ts:55]: Added

File: package.json
Comment (BarryNolte) [reply package.json:99]: Updated this PR

File: src/extension.ts
Comment (BarryNolte) [reply src/extension.ts:20]: Added an update handler that will reinitialize **ws** and not use the previously cached copy.

File: package.json
Comment (BarryNolte) [reply package.json:92]: Fixed

File: src/extension.ts
Comment (BarryNolte) [reply src/extension.ts:43]: Added change handler to avoid using cached values from the workspace.

Comment (gavin-ts) [reply src/extension.ts]: that's good but this one isn't about using cached values, it is to avoid the timer from rendering again if you have both settings on and save before the timer goes off.

File: .vscode/tasks.json
Comment (gavin-ts) [inline .vscode/tasks.json:38]: we removed the `watch-tests` script in `package.json` so we should also remove these right?

**Review by gavin-ts — APPROVED** [review]
nice! it is very nice to have the preview window with render right there while editing .d2 files 🚀

**Comment by alixander** [comment]
@BarryNolte this looks great, i think this is ready to merge, just wanted to double check with you it's set?

**Comment by BarryNolte** [comment]
> @BarryNolte this looks great, i think this is ready to merge, just wanted to double check with you it's set?

Yes, please. 

PR #16 has more features, some bug fixes and few better ways of doing things as I figured them out.  After that PR is merged, then we can drill down on bugs/features/enhancements with more targeted and smaller PR's.

**Comment by alixander** [comment]
@BarryNolte could you sign the commits please? I have it turned on as a branch protection org-wide. https://github.com/terrastruct/d2/pull/557#issuecomment-1367468730

**Comment by alixander** [comment]
@BarryNolte looks like it's been set up then, can you rewrite the commits? https://superuser.com/a/1123928 (except replace `development`)

**Comment by BarryNolte** [comment]
@alixander - I was able to partially get commits verified, but not all of them.  I had found that article you suggested and thought it would be unpleasant to use that technique, and it was.

Suggestion:  Create another PR with one signed commit that's the same as the last commit on this PR.  Merge it, and abandon this one.  I'll do the same for my other PR and then continue to sign each commit to avoid this in the future.

**Comment by BarryNolte** [comment]
Wow, that should have been easier...  See PR #19 

After that's merged, we can pull the plug on this PR.

**Comment by alixander** [comment]
sorry about the hoops @BarryNolte , ty for bearing with that, really excited to be merging all this for users (looks like the extension has been downloaded over 3k times now)

**Comment by BarryNolte** [comment]
I've had worse hoops 😄  I probably should have been signing my commits years ago.  Lesson learned and my love/hate relationship with git continues.  On to the next PR...

**Pull request closed without merging**
